### DISCUSSION TITLE
Is Address.Bytes() needed?

### DISCUSSION THREAD

**Discussion by ggwpez** [hyperledger-labs/go-perun#232] [Q&A]
A perun Address looks like this:  
```go
type Address interface {
	// Encoder should write an address that can be decoded with Backend.DecodeAddress.
	io.Encoder
	// Bytes should return the representation of the address as byte slice.
	Bytes() []byte
	…
```
We already had the problem that the `Bytes()` function does not work together with the decoding logic (that we now removed).  
Now i asked myself, why is it an encoder **and** has a `Bytes()` method?  
Should they do the same thing of not?  Simpler interfaces are always better IMHO.
@matthiasgeihs

**Comment by matthiasgeihs** [comment]
Can you check where `Address.Bytes()` is currently used?

**Reply by ggwpez** [reply]
The only occurrence except testing is in the kv-persister. Should be easy to kick out.

**Reply by matthiasgeihs** [reply]
I remember checking as well.

There it is used to generate a map key from an address object. It looks a little bit like a misusage to me, as what we ideally would want there is an Address.AsMapKey function. But is it cleaner to require an AsMapKey function? After all, it is common for an address type to have a Bytes method.

Something related I was wondering about: Why does the wallet.Address interface have a `Cmp` method? When checking, I saw that this method is actually never used by a pure wallet address, but only when a wallet address is used as a wire address. It seems like we introduced the method on the wallet.Address interface when it should actually only be on the wire address interface.

**Reply by ggwpez** [reply]
Ideed! `Cmp` is only used for the `wire.Address`, I will add an issue (ps: #234).  

Since the `AsMapKey` is only needed for the persistence, I think it should not bloat the `Address` interface.  
Just encoding the address and using that as key is fine AFAIK.
